pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/codestyle/ClassNamingConventionsRule.java

```diff
@@ -4,15 +4,144 @@
+
+        addRuleChainVisit(ASTClassOrInterfaceDeclaration.class);
+        addRuleChainVisit(ASTEnumDeclaration.class);
+        addRuleChainVisit(ASTAnnotationTypeDeclaration.class);
+    }
+
+
+    private void checkMatches(ASTAnyTypeDeclaration node, PropertyDescriptor<Pattern> regex, Object data) {
+        if (!getProperty(regex).matcher(node.getImage()).matches()) {
+            addViolation(data, node, new Object[]{
+                    isUtilityClass(node) ? "utility class" : node.getTypeKind().getPrintableName(),
+                    node.getImage(),
+                    getProperty(regex).toString(),
+            });
+        }
+    }
+
+
+    // This could probably be moved to ClassOrInterfaceDeclaration
+    // to share the implementation and be used from XPath
+    private boolean isUtilityClass(ASTAnyTypeDeclaration node) {
+        if (node.getTypeKind() != TypeKind.CLASS) {
+            return false;
+        }
+
+        // A class without declarations shouldn't be reported
+        boolean hasAny = false;
+
+        // we could probably enrich ASTAnyTypeBodyDeclaration
+        // with methods to know what kind of declaration it is
+        for (ASTAnyTypeBodyDeclaration decl : node.getDeclarations()) {
+            AccessNode accessNode = decl.getFirstChildOfType(AccessNode.class);
+            ASTInitializer initializer = decl.getFirstChildOfType(ASTInitializer.class);
+            if (accessNode != null && !(accessNode instanceof ASTConstructorDeclaration)) {
+                hasAny = true;
+                if (!accessNode.isStatic()) {
+                    return false;
+                }
+            }
+
+            if (initializer != null && !initializer.isStatic()) {
+                return false;
+            }
+        }
+
+        return hasAny;
+    }
+
+
+    @Override
     public Object visit(ASTClassOrInterfaceDeclaration node, Object data) {
-        if (Character.isLowerCase(node.getImage().charAt(0))) {
-            addViolation(data, node);
+
+        if (node.isAbstract()) {
+            checkMatches(node, ABSTRACT_CLASS_REGEX, data);
+        } else if (isUtilityClass(node)) {
+            checkMatches(node, UTILITY_CLASS_REGEX, data);
+        } else if (node.isInterface()) {
+            checkMatches(node, INTERFACE_REGEX, data);
+        } else {
+            checkMatches(node, CLASS_REGEX, data);
         }
+
         return data;
     }
+
+
+    @Override
+    public Object visit(ASTEnumDeclaration node, Object data) {
+        checkMatches(node, ENUMERATION_REGEX, data);
+        return data;
+    }
+
+
+    @Override
+    public Object visit(ASTAnnotationTypeDeclaration node, Object data) {
+        checkMatches(node, ANNOTATION_REGEX, data);
+        return data;
+    }
+
+
+    private static String toCamelCase(String name) {
+        StringBuilder sb = new StringBuilder();
+        boolean isFirst = true;
+        for (String word : name.trim().split("\\s++")) {
```

Comment:
what's the point of the second `+` ?

Reply:
It's probably useless in that situation since Matcher.find wouldn't backtrack. I've just grown a habit of making quantifiers possessive when possible